Support for Paraview/VTK XML (*.vtu) and Multiblock (*.vtm) visualization files

talbring opened the pull request:
## Proposed Changes
I added new output file formats. First of all support for the VTK XML file format ([see this pdf](https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf)) (file ending *.vtu). This adds support for possible compression, portable binary encoding, random access, big endian and little endian byte order, multiple file representation of piece data. I made this the new default format if `PARAVIEW` has been added to `OUTPUT_FILES`. The old *.vtk format can still be used with `PARAVIEW_LEGACY`.

However, the main reason for this new format is that it is required for new Multiblock file (*.vtm) output. This file is a simple XML file which defines a data hierarchy in order to represent individual parts of the domain (which are stored as *.vtu files).  With this it is possible to directly select and visualize the solution in individual zones and/or on boundaries. This new format can be enabled with `PARAVIEW_MULTIBLOCK` as value for `OUTPUT_FILES`. At the moment all boundaries are written, I might add an option to disable/enable some of them (maybe re-using the `MARKER_PLOTTING` option). In order not to cluster the working directory, the individual *.vtu files are placed in subfolders, since they are more or less opaque to the user. 

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

talbring commented:
The vtm file and the folder where the data is stored are now called like the config file (called "case name"). Futhermore the files inside of the folder are organized like the data. That means after a successful run, the working directory looks like that (omitting history files):
```
├── cht_2d_3cylinders
│   ├── zone_0
│   │   ├── cylinder_outer1.vtu
│   │   ├── cylinder_outer2.vtu
│   │   ├── cylinder_outer3.vtu
│   │   ├── farfield.vtu
│   │   └── Internal.vtu
│   ├── zone_1
│   │   ├── core1.vtu
│   │   ├── cylinder_inner1.vtu
│   │   └── Internal.vtu
│   ├── zone_2
│   │   ├── core2.vtu
│   │   ├── cylinder_inner2.vtu
│   │   └── Internal.vtu
│   └── zone_3
│       ├── core3.vtu
│       ├── cylinder_inner3.vtu
│       └── Internal.vtu
├── cht_2d_3cylinders.cfg
├── cht_2d_3cylinders.vtm
├── solid_cylinder1.cfg
├── solid_cylinder2.cfg
└── solid_cylinder3.cfg

```

And the structure in Paraview:

![cht_example](https://user-images.githubusercontent.com/4933860/71916713-d6f14d80-317e-11ea-9b78-8592d633a067.png)

pcarruscag commented on Common/src/config_structure.cpp:
```diff
@@ -9464,7 +9484,28 @@ void CConfig::SetMultizone(CConfig *driver_config, CConfig **config_container){
+    if (driver_config->GetVolumeOutputFiles()[iFiles] == PARAVIEW_MULTIBLOCK){
+      multiblockDriver = true;
+    }
+  }
```
You can do std::any_of if you want to be more c++11

pcarruscag commented on SU2_CFD/include/output/filewriter/CFEMDataSorter.hpp:
```diff
@@ -37,9 +37,9 @@ class CFEMDataSorter final: public CParallelDataSorter{
+   * \param[in] fieldNames - Vector containing the field names
    */
-  CFEMDataSorter(CConfig *config, CGeometry *geometry, unsigned short nFields);
+  CFEMDataSorter(CConfig *config, CGeometry *geometry, vector<string> fieldNames);
```
Please pass fieldNames by const&

pcarruscag commented on SU2_CFD/src/output/COutput.cpp:
```diff
@@ -424,27 +426,139 @@ void COutput::WriteToFile(CConfig *config, CGeometry *geometry, unsigned short f
+      
       volumeDataSorter->SortConnectivity(config, geometry, true);
-
+      
```
Remember to strip all blank spaces :)

TobiKattmann replied:
I wonder why CodeFactor did not complain

pcarruscag commented on SU2_CFD/include/output/filewriter/CParallelDataSorter.hpp:
```diff
@@ -232,4 +244,20 @@ class CParallelDataSorter{
+   */
+  vector<string> GetFieldNames(){
+    return fieldNames;
+  }
```
This can probably be returned by const&, it will still allow a local copy to be created if necessary.

pcarruscag commented on SU2_CFD/src/output/filewriter/CParaviewFileWriter.cpp:
```diff
@@ -267,19 +268,19 @@ void CParaviewFileWriter::Write_Data(){
+    string fieldname = fieldNames[iField];
 
     fieldname.erase(remove(fieldname.begin(), fieldname.end(), '"'), fieldname.end());
 
```
Could be my jet lagged eyes but it looks like this local manipulation of the name is not used?

pcarruscag replied:
Can you confirm or deny my observation?

talbring replied:
Its used a couple of lines below (line 287).

pcarruscag commented on SU2_CFD/src/output/filewriter/CParaviewVTMFileWriter.cpp:
```diff
@@ -0,0 +1,111 @@
+CParaviewVTMFileWriter::CParaviewVTMFileWriter(string fileName, string folderName, su2double time, 
+                                               unsigned short iZone, unsigned short nZone)
+  : CFileWriter(std::move(fileName), fileExt),
+    folderName(std::move(folderName)), iZone(iZone), nZone(nZone), curTime(time){
```
I would differentiate the argument / member names here (IDK if compilers complain or not, but just in case).

pcarruscag commented on SU2_CFD/src/output/filewriter/CParaviewVTMFileWriter.cpp:
```diff
@@ -0,0 +1,111 @@
+#else 
+    mkdir(this->folderName.c_str(), 0777); 
+    mkdir((this->folderName + "/zone_" + to_string(iZone)).c_str(), 0777);
+#endif
```
Should we have a "SU2_mkdir" in a toolbox to hide the windows/linux "hacks"?

pcarruscag commented on SU2_CFD/src/output/filewriter/CParaviewXMLFileWriter.cpp:
```diff
@@ -0,0 +1,682 @@
+    default:
+      SU2_MPI::Error("Unknown Type", CURRENT_FUNCTION);
+      break;
+  }
```
This is done also in AddDataArray, maybe it could be in a function to make maintenance easier if more types are added?

pcarruscag commented on SU2_CFD/src/output/filewriter/CSurfaceFVMDataSorter.cpp:
```diff
@@ -1119,7 +1133,31 @@ void CSurfaceFVMDataSorter::SortConnectivity(CConfig *config, CGeometry *geometr
 }
 
-void CSurfaceFVMDataSorter::SortSurfaceConnectivity(CConfig *config, CGeometry *geometry, unsigned short Elem_Type) {
+void CSurfaceFVMDataSorter::SortConnectivity(CConfig *config, CGeometry *geometry, vector<string> markerList) {
```
Another vector that should be passed by const& here and in SortSurfaceConnectivity.

talbring commented:
Thanks @pcarruscag for your comments. As always very valuable! I think I have addressed all of them.

Some other things I have changed:

- The history files per zone are disabled by default when running a multizone problem
- The multizone history file is now named after the config file (i.e. case name).

TobiKattmann commented on SU2_CFD/src/output/COutput.cpp:
```diff
@@ -654,19 +787,19 @@ void COutput::PrintConvergenceSummary(){
   for (unsigned short iField_Conv = 0; iField_Conv < convFields.size(); iField_Conv++){
     const string &convField = convFields[iField_Conv];
-    if (historyOutput_Map[convField].fieldType == HistoryFieldType::COEFFICIENT) {
+    if (historyOutput_Map.at(convField).fieldType == HistoryFieldType::COEFFICIENT) {
```
out of interest: Why do you prefer .at over [] here? Is there something more than the out_of_bounds check and thrown exception?

talbring replied:
The [] operator for a map does not check whether that element exists. If it doesn't exist, it will be implicitly created, which can lead to strange behavior. Furthermore, you cannot use it in a const member function.

economon replied:
Any performance difference? I know with .at() there is some overhead due to the extra checks, so for large data sets it may not be ideal. It shouldn't be noticeable for just a handful of checks though

talbring replied:
There shouldn't be any difference. Because the code for operator[] and at() is essentially the same. The only difference is that at() throws an exception and operator[] adds a new item in case the element cannot be found.

TobiKattmann replied:
For completeness on this [] vs .at() : .at() is a const operation on the data whilst [] isn't because [] creates the entry if it is not present in the map, as Tim states above. And now that we are adding more const's to the code and therefore const-correctness could become tricky one day -> .at() has your back.

talbring commented on Common/include/config_structure.hpp:
```diff
@@ -9354,6 +9355,12 @@ class CConfig {
+   * \brief Get the name of the current case
+   * \return the case name
+   */
+  string GetCaseName();
```
yeah probably. Didn't think about it since no other routine in CConfig has it so far. I can change that. We have remember to do it when restructuring.

talbring commented on SU2_CFD/include/output/filewriter/CParaviewVTMFileWriter.hpp:
```diff
@@ -0,0 +1,131 @@
+  /*!
+   * \brief String stream that stores the content of the vtm file
+   */
+  stringstream output;
```
Do we have to save space ? :D I guess that this is a personal preference.

TobiKattmann replied:
real estate's expensive these days 🏡 
I prefer the compact list.

talbring commented on SU2_CFD/src/output/filewriter/CParaviewXMLFileWriter.cpp:
```diff
@@ -0,0 +1,613 @@
+    SU2_MPI::Error("Connectivity must be sorted.", CURRENT_FUNCTION);
+  }
+
+  /*--- We always have 3 coords, independent of the actual value of nDim ---*/
```
Yep, it always assumes three dimensions

talbring commented on SU2_CFD/src/output/filewriter/CSurfaceFVMDataSorter.cpp:
```diff
@@ -1115,11 +1129,36 @@ void CSurfaceFVMDataSorter::SortConnectivity(CConfig *config, CGeometry *geometr
+  nGlobal_Elem_Par   = nTotal_Surf_Elem;
+#else
+  SU2_MPI::Allreduce(&nTotal_Surf_Elem, &nGlobal_Elem_Par, 1, MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+#endif
```
Well if I start with that now, I'll have to change it everywhere. Let's save that for a separate PR.

TobiKattmann replied:
Sure, it would be better to get rid of a lot of unnecessary preprocessor statements for MPI calls in a separate PR but that shouldn't be a free ticket introduce these things in new code. In COutput.cpp +559 you also use Allreduce without that "padding".

pcarruscag replied:
+1 on this one, I've been replacing those things wherever I see them, better do something than do nothing.

economon replied:
👍, yep, a lot of this is holdover from before we had the wrapper. Best practice is to hide the ifdefs within mpi_structure whenever possible!

economon replied:
I agree with just doing them as you go

talbring commented on SU2_CFD/include/output/filewriter/CParaviewXMLFileWriter.hpp:
```diff
@@ -0,0 +1,157 @@
+
+class CParaviewXMLFileWriter final: public CFileWriter{
+
+private:
```
I like to have all variable declarations first and then all function declarations.

TobiKattmann replied:
I ended up doing the same for the STL-PR, as it is quite a useful separation

economon commented on SU2_CFD/src/output/filewriter/CParaviewXMLFileWriter.cpp:
```diff
@@ -0,0 +1,613 @@
+
+}
+
+void CParaviewXMLFileWriter::WriteDataArray(void* data, VTKDatatype type, unsigned long arraySize,
```
I like how you broke these methods calling the MPI IO out. Seems to me that some of them could potentially be moved up to the parent class to save duplication. We use similar patterns for the writing of contiguous chunks of binary data (restart, ParaView writers, etc.). Is this possible without too much extra work?

talbring commented:
Found a bug where I get a deadlock for a big case while writing. Have to look into it to see whether its a problem on our cluster or in the code. If someone could test the vtu/vtm output on a big case (> 32 cores), that would be also very helpful.

TobiKattmann commented:
> 
> 
> Found a bug where I get a deadlock for a big case while writing. Have to look into it to see whether its a problem on our cluster or in the code. If someone could test the vtu/vtm output on a big case (> 32 cores), that would be also very helpful.

I don't have these problems. I ran the coupled_cht/incomp_2d case on 32 and 56 cores (28 cores/node) without any issues.

talbring commented:
Problems are solved. PR is ready from my side.

pcarruscag commented on SU2_CFD/src/output/filewriter/CParallelDataSorter.cpp:
```diff
@@ -436,3 +410,59 @@ unsigned long CParallelDataSorter::GetElem_Connectivity(GEO_TYPE type, unsigned
+  SU2_MPI::Allreduce(nLocalPerElem.data(), nGlobalPerElem.data(), N_ELEM_TYPES, MPI_UNSIGNED_LONG, MPI_SUM, MPI_COMM_WORLD);
+  
+  nGlobalElem = std::accumulate(nGlobalPerElem.begin(), nGlobalPerElem.end(), 0); 
+  nLocalElem  = std::accumulate(nLocalPerElem.begin(), nLocalPerElem.end(), 0);
```
Ah yes... me likes.

pcarruscag commented on SU2_CFD/src/output/filewriter/CParallelFileWriter.cpp:
```diff
@@ -28,24 +28,249 @@
+
+  startTime = su2double(clock())/su2double(CLOCKS_PER_SEC);
+  
+  unsigned long bytesWritten;
```
I have my eye on moving this timing stuff to the wrapper at some point.

pcarruscag commented on SU2_CFD/src/output/filewriter/CParaviewVTMFileWriter.cpp:
```diff
@@ -0,0 +1,119 @@
+#else
+    mkdir(this->folderName.c_str(), 0777);
+    mkdir((this->folderName + "/zone_" + to_string(valiZone)).c_str(), 0777);
+#endif
```
I did not get enough sleep but I don't think you have enough "{}"

talbring replied:
oh yes you are right :D fortunately if the folders already exists, _mkdir() just does nothing. But I am gonna change this.

pcarruscag commented on SU2_DOT/src/meson.build:
```diff
@@ -1,28 +1,31 @@
+                                				 'output/filewriter/CParallelFileWriter.cpp',
+                                				 'output/filewriter/CFEMDataSorter.cpp',
+                                				 'output/filewriter/CSurfaceFEMDataSorter.cpp',
+                                				 'output/filewriter/CFVMDataSorter.cpp',
```
Please align these.

pcarruscag commented on SU2_SOL/src/meson.build:
```diff
@@ -21,6 +21,8 @@ if get_option('enable-normal')
                                         'output/filewriter/CSU2MeshFileWriter.cpp',
+                                        'output/filewriter/CParaviewXMLFileWriter.cpp',
+                                        'output/filewriter/CParaviewVTMFileWriter.cpp',
                                         'variables/CBaselineVariable.cpp',
```
I might have missed it, but can you add these to the old build system too?

TobiKattmann replied:
I think this is not addressed yet.

TobiKattmann commented on SU2_CFD/include/output/filewriter/CParallelDataSorter.hpp:
```diff
@@ -122,36 +135,97 @@ class CParallelDataSorter{
    * \return local number of points.
    */
-  unsigned long GetnPoints(){return nParallel_Poin;}
+  unsigned long GetnPoints() const {return nLocalPoint;}
```
bueno 💃

TobiKattmann commented on SU2_CFD/include/output/filewriter/CSurfaceFEMDataSorter.hpp:
```diff
@@ -31,21 +31,20 @@
+  CFEMDataSorter* volumeSorter;                  //!< Pointer to the volume sorter instance
    //! Structure to map the local sorted point ID to the global point ID
   std::vector<unsigned long> globalSurfaceDOFIDs;
   vector<unsigned long> nSurfaceDOFsRanks;
```
bene

pcarruscag commented on SU2_CFD/obj/Makefile.am:
```diff
@@ -87,6 +87,8 @@ libSU2Core_sources = ../src/definition_structure.cpp \
   ../src/output/filewriter/CParallelFileWriter.cpp \
   ../src/output/filewriter/CParaviewBinaryFileWriter.cpp \
+  ../src/output/filewriter/CParaviewXMLFileWriter.cpp \
+  ../src/output/filewriter/CParaviewVTMFileWriter.cpp \
```
Thanks LGTM
